app/tests/fixtures/open-secret/fixture.ts

```diff
@@ -0,0 +1,186 @@
+import { decode } from '@stablelib/base64';
+import type { Serializable } from 'playwright-core/types/structs';
+import { openSecretEncryption } from '~/tests/fixtures/open-secret/encryption';
+import { session } from '~/tests/mocks/open-secret';
+
+type OpenSecretApiMockOptions<T extends Serializable> = {
+  /**
+   * Url of the endpoint. Can be full url or just a path
+   */
+  url: string;
+
+  /**
+   * Response headers. Header values will be converted to a string.
+   */
+  headers?: { [key: string]: string };
+
+  /**
+   * JSON response.
+   */
+  responseData: T;
+
+  /**
+   * Response status code, defaults to `200`.
+   */
+  status?: number;
+
+  /**
+   * How often a mock should be used. By default, it will be used every time.
+   */
+  times?: number;
+};
+
+type OpenSecretApiMock = {
+  /**
+   * Sets up open secret api mock with raw response data.
+   */
+  setup: <T = unknown>(options: OpenSecretApiMockOptions<T>) => Promise<void>;
+  /**
+   * Sets up open secret api mock with encrypted response data.
+   */
+  setupEncrypted: <T = unknown>(
+    options: OpenSecretApiMockOptions<T>,
+  ) => Promise<void>;
+};
+
+export type OpenSecretFixture = {
+  /**
+   * Open secret API mock. Use to mock the Open Secret responses.
+   */
+  openSecretApiMock: OpenSecretApiMock;
+  /**
+   * Presets Open Secret session before starting a test.
+   * True by default. Set to false if you want to test "session handshake"
+   */
+  presetOpenSecretSession?: boolean;
+};
+
+type MockData = {
+  config: OpenSecretApiMockOptions<unknown>;
+  numOfInvocations: number;
+};
+
+const openSecretBaseUrl = 'https://preview-enclave.opensecret.cloud';
+
+export const test = base.extend<OpenSecretFixture>({
+  presetOpenSecretSession: [true, { option: true }],
+  openSecretApiMock: async (
+    { page, baseURL, presetOpenSecretSession },
+    use,
+  ) => {
+    const mocksMap = new Map<string, MockData>();
+    let openSecretRequestsCounter = 0;
+
+    page.on('request', (request) => {
+      if (request.url().startsWith(openSecretBaseUrl)) {
+        openSecretRequestsCounter++;
+      }
+    });
+
+    if (presetOpenSecretSession) {
+      await page.addInitScript(
+        async ({ baseURL, session }) => {
+          if (
+            window.location.origin === baseURL &&
+            !window.sessionStorage.getItem('sessionId') &&
+            !window.sessionStorage.getItem('sessionKey')
+          ) {
+            window.sessionStorage.setItem('sessionId', session.id);
+            window.sessionStorage.setItem('sessionKey', session.key);
+          }
+        },
+        { baseURL, session },
+      );
+    }
+
+    const setup: OpenSecretApiMock['setup'] = async (options) => {
+      const optionsStr = JSON.stringify(options);
+      const key = crypto.createHash('sha256').update(optionsStr).digest('hex');
+      const mockData = mocksMap.get(key);
+      if (!mockData) {
+        mocksMap.set(key, { config: options, numOfInvocations: 0 });
```

Comment:
This seems like it is so that if a test sets up a mock that  has already been setup, then we just reuse the original. Why would the same mock be setup twice?

Reply:
when you call `page.route` in playwright twice with the same url the last invocation wins and previous one is ignored/cancelled. this is why I had to come up with this approach where I take other options into account as well and count number of invocations based on that.

if the subsequent call to the mock setup has all the same options, I didn't want to reset the existing numOfInvoications count which is why I don't set the data in mocksMap if it already exists.

the reason why someone would call it twice with the same options would most probably be a mistake and this approach will basically ignore it. the other reason I could think of is if you would want to increase the num of times. e.g. you setup the mock with times 1 and now you want to add one more. to support that we would need to update mocks map data to sum the num of times for each invocations but I am not sure if that is even a valid use case so didn't want to implement it unless we see a need in practice